src/array/struct_/mutable.rs

```diff
@@ -0,0 +1,207 @@
+use std::sync::Arc;
+
+use crate::{
+    array::{Array, MutableArray},
+    bitmap::MutableBitmap,
+    datatypes::DataType,
+};
+
+use super::StructArray;
+
+/// Converting a [`MutableStructArray`] into a [`StructArray`] is `O(1)`.
+#[derive(Debug)]
+pub struct MutableStructArray {
+    data_type: DataType,
+    values: Vec<Box<dyn MutableArray>>,
+    validity: Option<MutableBitmap>,
+}
+
+impl From<MutableStructArray> for StructArray {
+    fn from(other: MutableStructArray) -> Self {
+        let validity = if other.validity.as_ref().map(|x| x.unset_bits()).unwrap_or(0) > 0 {
+            other.validity.map(|x| x.into())
+        } else {
+            None
+        };
+
+        StructArray::from_data(
+            other.data_type,
+            other.values.into_iter().map(|mut v| v.as_box()).collect(),
+            validity,
```

Comment:
Even though "equal child lengths" can't truly be an invariant due to the need to push children one at a time, it still feels like an invariant to me. As a consumer of this API I'd prefer it to panic like `push` does, because I can't think of when this would happen outside of programmer error. But I can switch to `TryFrom` if you prefer.

Reply:
you are right - this is a programatic error that the user can't recover from 👍